blog-posts/cloudfront-cache-efficiency.md

```diff
@@ -0,0 +1,154 @@
+requests. As caching times in the `Cache-Control` header are relative to the response `date` header, only the first 
+request by the browser get's cached for the `max-age` time. 
+
+> CDNs usually don't cache the `date` header, it's only CloudFront that does it.
+
+There is no easy way to fix this within some CloudFront settings. Only recently it is possible to solve this problem at
+all: With a custom CloudFront Function. CloudFront Functions are small Javascript functions that are executed with every
+request. You can use them to do simple changes to the request or response object. Read more about the possibilities and
+limit in the official AWS docs
+here: [https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/functions-javascript-runtime-features.html](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/functions-javascript-runtime-features.html)
+
+We can use those functions to update the `date` header with every request.
+
+```js
+function handler(event) {
+  var response = event.response;
+
+  response.headers['date'] = {
+    value: new Date().toUTCString()
+  };
+  delete response.headers['age'];
+
+  return response;
+}
+```
+
+This code updates the `date` header with the current time. It also deletes the `age` header, as it turns out that
+CloudFront is also setting it. The [age header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Age) is used
+to tell how long the cached object is in the cache. If it is set, the browser will compare it against the `max-age` time
+and decide to not cache the object. Which is not what we want.
+
+After deploying the CloudFront function, let’s check the browser console:
+
+![Requests are cached by the browser](images/cloudfront-cache-efficiency/2-improved.png)
+
+What do we see here:
+
+1. The first request hits Lambda, so it takes around 700ms.
+2. The requests within 10 seconds are cached. After that, one request is done to CloudFront. This request takes around 30ms
+   and is cached again for 10 seconds by the browser. CloudFront returns the full response.
+3. Only after 1 minute, the requests hits Lambda again and gets then cached again by CloudFront.
+
+This is the behavior we want! Nice!
+
+Next, let’s see how we can return `304` if we hit CloudFront and the response hasn't changed since the last request.
+
+## Return 304 if the object hasn’t changed in CloudFront
+
+CloudFront always returns `200` with the full response body. This results in a lot of unnecessary data send between the
+browser and CloudFront. To fix this, we need to set the `ETag` header in our Lambda function.
+
+The `ETag` is a hash for the response body. The browser will send this hash with alongside with all upcoming requests to
+the same resource. The server can then decide if the resource has changed and either return `200` with a new resource
+or `304` .
+
+Luckily, when the `ETag` header is set, CloudFront will handle `200` and `304` automatically. The `ETag` need to be
```

Comment:
Would it make a difference if we set the ETag on our side without tinkering with the CDN? Or would cloudfront ignore that ?

Reply:
We already set the `ETag` on our side. CloudFront uses this header then without any additional steps that we need to do. I added an additional sentence to make it more clear.